include/zephyr/drivers/stepper.h

```diff
@@ -0,0 +1,425 @@
+	STEPPER_MICRO_STEP_256,
+};
+
+/**
+ * @brief Stepper Motor direction options
+ */
+enum stepper_direction {
+	/** Positive direction */
+	STEPPER_DIRECTION_POSITIVE = 0,
+	/** Negative direction */
+	STEPPER_DIRECTION_NEGATIVE,
+};
+
+/**
+ * @brief Stepper Motor signal results
+ */
+enum stepper_signal_result {
+	/** Target position reached */
+	STEPPER_SIGNAL_TARGET_POSITION_REACHED = 0,
+};
+
+/**
+ * @cond INTERNAL_HIDDEN
+ *
+ * Stepper motor controller driver API definition and system call entry points.
+ *
+ * (Internal use only.)
+ */
+
+/**
+ * @brief enable or disable the stepper motor controller.
+ *
+ * @see stepper_enable() for details.
+ */
+typedef int (*stepper_enable_t)(const struct device *dev, bool enable);
+
+/**
+ * @brief Move the stepper motor by a given number of micro_steps.
+ *
+ * @see stepper_move() for details.
+ */
+typedef int (*stepper_move_t)(const struct device *dev, int32_t micro_steps,
+				  struct k_poll_signal *async);
+
+/**
+ * @brief Set the max velocity in micro_steps per seconds.
+ *
+ * @see stepper_set_max_velocity() for details.
+ */
+typedef int (*stepper_set_max_velocity_t)(const struct device *dev,
+					      uint32_t micro_steps_per_second);
+
+/**
+ * @brief Set the micro-step resolution
+ *
+ * @see stepper_set_micro_step_res() for details.
+ */
+typedef int (*stepper_set_micro_step_res_t)(const struct device *dev,
+						enum micro_step_resolution resolution);
+
+/**
+ * @brief Get the micro-step resolution
+ *
+ * @see stepper_get_micro_step_res() for details.
+ */
+typedef int (*stepper_get_micro_step_res_t)(const struct device *dev,
+						enum micro_step_resolution *resolution);
+/**
+ * @brief Set the actual a.k.a reference position of the stepper
+ *
+ * @see stepper_set_actual_position() for details.
+ */
+typedef int (*stepper_set_actual_position_t)(const struct device *dev, const int32_t value);
+
+/**
+ * @brief Get the actual a.k.a reference position of the stepper
+ *
+ * @see stepper_get_actual_position() for details.
+ */
+typedef int (*stepper_get_actual_position_t)(const struct device *dev, int32_t *value);
+
+/**
+ * @brief Set the absolute target position of the stepper
+ *
+ * @see stepper_set_target_position() for details.
+ */
+typedef int (*stepper_set_target_position_t)(const struct device *dev, const int32_t value,
+						 struct k_poll_signal *async);
+
+/**
+ * @brief Is the target position fo the stepper reached
+ *
+ * @see stepper_is_moving() for details.
+ */
+typedef int (*stepper_is_moving_t)(const struct device *dev, bool *is_moving);
+
+/**
+ * @brief Enable constant velocity mode for the stepper with a given velocity
+ *
+ * @see stepper_enable_constant_velocity_mode() for details.
+ */
+typedef int (*stepper_enable_constant_velocity_mode_t)(const struct device *dev,
+							   const enum stepper_direction direction,
+							   const uint32_t value);
+
+/**
+ * @brief Stepper Motor Controller API
+ */
+__subsystem struct stepper_api {
```

Comment:
I disagree with a API that will potentially grow up with function callbacks (system calls !!!) – use simple setter/getter pattern for parameters or attributes that have to write or read. Also avoid context related status functions (e.g. "is running") that is on interesting information but not the only one – or not?

Reply:
Hi @rexut, 

the idea of using attributes was actually reviewed once in the past. I think what you intend kind of looked like this: https://github.com/zephyrproject-rtos/zephyr/pull/68774#discussion_r1551462543 as @fabiobaltieri had suggested at that time. 

However, over the period of time a consensus was built that driver specific functionalities such as Stallguard from trinamic need to be made available through a public header file and not mixed with the general stepper api as such. Ramp parameters are also again highly `vendor-specific` and there is no one common way to package them. What could also be done is using `void *config` in order to pack driver specific configuration in a struct. 

Hence, i would like to have a basic structure providing a set of generalised functions as of now and then build on top of it.

Reply:
I see, but disagree this practice. As I say, API functions can be system calls and system calls are limited on some arches, especially when you enable the Zephyr user space feature. The goal of APIs should be to reduce the number of calls. In Zephyr we can't let explode the interfaces like in C++ or Java.

Reply:
> However, over the period of time a consensus was built that driver specific functionalities such as Stallguard from trinamic need to be made available through a public header file and not mixed with the general stepper api as such

Time to change the mind set, away from a "Stepper Control API". That would be step two. Start with the basics and see what will be missed in the future, when such kind of functionality will be integrated in any upper layer component and how the related information flow from upper layer down to the hardware (through this API) can be supported.

> Ramp parameters are also again highly vendor-specific

Also a good point, why this API should never handle anything of "control", "management", "sequencing", or whatever feature / behavior users need is.

All this lead to a very very simple and handy set of API functions:

1. power stages on –> motor in active break or similar
2. power stages off -> motor w/o current
3. motor action -> movement propagation for different operation modes (distance, speed and conceivably also acceleration)

So, you see. Three functions are enough for the basics.

With more interaction with upper level functions or application code we will get a few more API functions:

1. **get & set active movement parameters** -> only a few, e.g. velocity set when and active movement is already in running either in positioning or velocity mode – **I can't think of any more parameters so far. And that is the key reason why the getter/setter pattern is definitely the right approach for this API.**
2. fetch motor driver capabilities -> such as: steps per revolution (reflects the minimal angle step), max. speed, and certainly many more – all that are important values for the positioning or navigation arithmetic in an upper level controller or management component
3. register and remove callbacks for events -> that is important and skips the currently intended polling architecture (see your `is_moving()`)